python3/tests/test_parser.py

```diff
@@ -283,3 +283,123 @@ def test_parse_oscillator(self):
         assert osc['frequency'] == 16 * mega
         assert osc['tolerance'] == 1
         assert osc['capacitance'] == 12 * pico
+
+
+class ParseDiodeTests(unittest.TestCase):
+    def test_parse_code(self):
+        parse = parser('electro_grammar', key=False)
+        d = parse('1N4148')
+        assert d['type'] == 'diode'
+        assert d['code'] == '1N4148'
+
+    def test_parse_signal(self):
+        parse = parser('electro_grammar', key=False)
+        d = parse('sig')
+        assert d['type'] == 'diode'
+        assert d['dtype'] == 'signal'
+
+        d = parse('sig 1N4148')
+        assert d['type'] == 'diode'
+        assert d['dtype'] == 'signal'
+        assert d['code'] == '1N4148'
+
+        d = parse('1N4148 sig')
+        assert d['type'] == 'diode'
+        assert d['dtype'] == 'signal'
+        assert d['code'] == '1N4148'
+
+    def test_parse_rectifier(self):
+        parse = parser('electro_grammar', key=False)
+        d = parse('rect')
+        assert d['type'] == 'diode'
+        assert d['dtype'] == 'rectifier'
+
+        d = parse('rect 1N4007')
+        assert d['type'] == 'diode'
+        assert d['dtype'] == 'rectifier'
+        assert d['code'] == '1N4007'
+
+        d = parse('1N4007 rect')
+        assert d['type'] == 'diode'
+        assert d['dtype'] == 'rectifier'
+        assert d['code'] == '1N4007'
+
+    def test_parse_schottky(self):
+        parse = parser('electro_grammar', key=False)
+        d = parse('sk')
+        assert d['type'] == 'diode'
+        assert d['dtype'] == 'schottky'
+
+        d = parse('sk BAT54S')
+        assert d['type'] == 'diode'
+        assert d['dtype'] == 'schottky'
+        assert d['code'] == 'BAT54S'
```

Comment:
What's the `code`? Is it the manufacturer part number?

Reply:
yes, we can change it to mpn. the difference is only that for semiconductors the mpn has a structure, for arbitrary ic's it does not.